### ISSUE TITLE
Record constructors should not trigger "Constructor without 'inherited' statement" rule

### ISSUE THREAD

**Issue by Teloah** [fabriciocolombo/sonar-delphi#17]
As records cannot have virtual constructors, `"Constructor without 'inherited' statement"` rule should not apply to them.

``` delphi
unit RecordConstructor;

interface

type
  TDummyRec = record
    FData : Integer;
    constructor Create(aData : Integer);
  end;

implementation

constructor TDummyRec.Create(aData : Integer);
begin
  FData := aData;
end;

end.
```

**Comment by fabriciocolombo** [comment]
I Agree

**Comment by Teloah** [comment]
I'm starting to learn how all of this stuff works, so I can take a shot at this and possibly something out of TODO list.

What I figured out so far:
- ANTLR creates AST based on the grammar in [Delphi.g](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/antlr3/org/sonar/plugins/delphi/antlr/Delphi.g) file. Judging by directory name `antlr3`, I should read [this introduction](https://theantlrguy.atlassian.net/wiki/display/ANTLR3/Five+minute+introduction+to+ANTLR+3). By the way, what's your preference in "spaces vs tabs" holywar? That file seems [quite inconsistent](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/antlr3/org/sonar/plugins/delphi/antlr/Delphi.g#L760) in that regard.
- [PMD rules](https://pmd.github.io/pmd-5.4.0/customizing/howtowritearule.html) are visitors that traverse AST, check its nodes and add violations where needed. I have to update [NoInheritedStatementRule.java](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/java/org/sonar/plugins/delphi/pmd/rules/NoInheritedStatementRule.java). Probably this file will continue to serve destructors, but constuctors will require a different file, because differences will be quite significant.
- It seems that [the current test for inherited constructors](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/pmd/DelphiPmdSensorTest.java#L155) is not the most maintainable test, because check `assertThat("number of issues", issues, hasSize(ruleData.length));` is already commented out and any update of [pmd.pas](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/resources/org/sonar/plugins/delphi/PMDTest/pmd.pas) file requires update of all those line numbers in the test. I think I'll create dedicated tests.

My problems so far:
- Eclipse ignores(?) `@SneakyThrows` attributes and shows 3 errors about unhandled exceptions. It is not a showstopper, because tests still run and `mvn package` succesfully builds the `jar` file, but it is annoying anyway.
-  2 tests fail with `java.lang.NoSuchMethodError`: `SurefireSensorTest.analyzeTest` and `SurefireSensorTest.analyzeTestUsingDefaultSurefireReportsPath`. Again, it is not a showstopper, but I'd like to see that green bar for the whole suite.
- How to see the whole generated AST? Is there an easy way to output it to the console or a file?

So I don't promise any quick results, but I will continue to play with this. Any corrections, tips and suggestions, so that I do not dig in a completely wrong direction?

**Comment by fabriciocolombo** [comment]
Hi @Teloah, you are in the right direction. Your feedback are very appreciated, and showed me some points to improve in a way to facilitate the contribution.

> ANTLR creates AST based on the grammar in [Delphi.g](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/antlr3/org/sonar/plugins/delphi/antlr/Delphi.g) file. Judging by directory name `antlr3`, I should read [this introduction](https://theantlrguy.atlassian.net/wiki/display/ANTLR3/Five+minute+introduction+to+ANTLR+3). By the way, what's your preference in "spaces vs tabs" holywar? That file seems [quite inconsistent](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/antlr3/org/sonar/plugins/delphi/antlr/Delphi.g#L760) in that regard.

I prefer `spaces`, as configured to format the source code. I will correct the formatting now.

> [PMD rules](https://pmd.github.io/pmd-5.4.0/customizing/howtowritearule.html) are visitors that traverse AST, check its nodes and add violations where needed. I have to update [NoInheritedStatementRule.java](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/java/org/sonar/plugins/delphi/pmd/rules/NoInheritedStatementRule.java). Probably this file will continue to serve destructors, but constuctors will require a different file, because differences will be quite significant.

Some rules are implemented by the [same class](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/main/resources/org/sonar/plugins/delphi/pmd/rules.xml#L409-L427) and use properties to configure what to do. I think this it's a bad practice, and i prefer create specific classes for Constructors and Destructors, then we could remove the properties from rules.xml.

> It seems that [the current test for inherited constructors](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/pmd/DelphiPmdSensorTest.java#L155) is not the most maintainable test, because check `assertThat("number of issues", issues, hasSize(ruleData.length));` is already commented out and any update of [pmd.pas](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/resources/org/sonar/plugins/delphi/PMDTest/pmd.pas) file requires update of all those line numbers in the test. I think I'll create dedicated tests.

The commented lines on [DelphiPmdSensorTest](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/pmd/DelphiPmdSensorTest.java) are my bad. By the way, this tests are like `deprecated`, we should create specific tests for each rule. The [EmptyBeginStatementTest.java](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/pmd/EmptyBeginStatementTest.java) it is a good example of how these tests look like. Basicaly extends from [BasePmdRuleTest  class](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/pmd/BasePmdRuleTest.java), fill the "builder" with the source code under test, call the analyze method and verify the generated issues.
I will create the dedicated test, than you can improve it to cover this issue.

> My problems so far:
>  Eclipse ignores(?) `@SneakyThrows` attributes and shows 3 errors about unhandled exceptions. It is not a showstopper, because tests still run and `mvn package` succesfully builds the `jar` file, but it is annoying anyway.

It's because project  [lombok](https://github.com/fabriciocolombo/sonar-delphi/blob/master/pom.xml#L237-L242) should be installed in the IDE to enable the use of annotation processing. I prefer remove this dependency and keep the project more simple.

>  2 tests fail with `java.lang.NoSuchMethodError`: `SurefireSensorTest.analyzeTest` and `SurefireSensorTest.analyzeTestUsingDefaultSurefireReportsPath`. Again, it is not a showstopper, but I'd like to see that green bar for the whole suite.

I can't reproduce this behaviour. Here and on https://travis-ci.org/fabriciocolombo/sonar-delphi,  all the tests are passing successfully. Give me more details about your environament: OS, IDE, JDK Version, etc.

> How to see the whole generated AST? Is there an easy way to output it to the console or a file?

The [GrammarTest](https://github.com/fabriciocolombo/sonar-delphi/blob/master/src/test/java/org/sonar/plugins/delphi/antlr/GrammarTest.java) it's for this purpose. It test give a input source file and generate the AST as a XML File.

Thank you for your investigation. 
Wait for my fixes, than you can continue to play with this.

**Comment by fabriciocolombo** [comment]
Done. I created specific classes for rule and tests. Add a test case and try fix it.
Let me know if you need some help or have any doubt.

**Comment by Teloah** [comment]
I fixed the problem with those failing SurefireSensor tests. I noticed the following lines in the console when running the failing tests:

```
SLF4J: The requested version 1.6 by your slf4j binding is not compatible with [1.5.5, 1.5.6]
SLF4J: See http://www.slf4j.org/codes.html#version_mismatch for further details.
```

Stack trace also mentioned slf4j:

```
java.lang.NoSuchMethodError: org.slf4j.helpers.MessageFormatter.arrayFormat(Ljava/lang/String;[Ljava/lang/Object;)Lorg/slf4j/helpers/FormattingTuple;
    at ch.qos.logback.classic.spi.LoggingEvent.getFormattedMessage(LoggingEvent.java:298)
    at ch.qos.logback.classic.spi.LoggingEvent.prepareForDeferredProcessing(LoggingEvent.java:208)
```

Eclipse's "Dependency Hierarchy" tab for `pom.xml` shows that slf4j is used by 2 components:
- `sonar-plugin-api` requires version `1.7.5` of `slf4j-api` and `jcl-over-slf4j`;
- `sslr-squid-bridge` requires version `1.5.6` of those libraries.

And somehow `mvn eclipse:eclipse` had put versions `1.5.6` in `.classpath`. I updated them to `1.7.5` and now all tests pass.

As I had this problem on 2 different clean machines, I think others could have the same problem too. But I have no idea how to force maven to choose the right versions.

So I have no more technical problems and will continue to play with all this stuff. I noticed that this rule incorrectly marks chained constructors for classes too:

```
constructor TSomeClass.Create(aParam : Integer);
begin
  inherited;
  FField := aParam;
end;

constructor TSomeClass.Create;
// this constructor should not be marked as a violation
begin
  Create(0);
end;
```

But first I will try to fix the subject of this issue.

**Comment by Teloah** [comment]
While trying to figure out how to fix this issue, [I added a test and a few debug `System.out.println` statements](https://github.com/Teloah/sonar-delphi/commit/8e6e9b7cbf9bd4e1eabd6d3b448eea0dc75a2c7c) to see in which order and which nodes are passed to `NoInheritedStatementRule.visit`.
I was planning to do something like this:
- catch all `constructor` nodes in the interface section; by querying their `TkNewType` ancestor's children figure out if current type is a class or a record and remember that along with type's name;
- catch all `constructor` nodes in the implementation section and by name figure out if it is a record or a class.

But the output was surprising: no nodes from `record` declaration are passed in:

``` delphi
Node        > [unit]
Node        > [Unit1]
Node        > [interface]
Section     > (interface (type (TkNewType (TTestRecord record FData (TkVariableType Integer) (constructor (TkFunctionName Create) (TkFunctionArgs (TkVariableIdents aData) (TkVariableType Integer))) end))))
Node        > [type]
Node        > [TkNewType]
Node        > [end]
Node        > [implementation]
Section     > (implementation (constructor (TkFunctionName TTestRecord . Create) (TkFunctionArgs (TkVariableIdents aData) (TkVariableType Integer))) (begin inherited ; FData := aData ; end))
Node        > [constructor]
Constructor > (constructor (TkFunctionName TTestRecord . Create) (TkFunctionArgs (TkVariableIdents aData) (TkVariableType Integer)))
Node        > [TkFunctionName]
Node        > [TTestRecord]
Node        > [.]
Node        > [Create]
Node        > [TkFunctionArgs]
Node        > [TkVariableIdents]
Node        > [aData]
Node        > [TkVariableType]
Node        > [Integer]
Node        > [begin]
Node        > [inherited]
Node        > [;]
Node        > [FData]
Node        > [:=]
Node        > [aData]
Node        > [;]
Node        > [end]
Node        > [end]
Node        > [.]
```

That is not the case for `class` types:

``` delphi
Node        > [unit]
Node        > [Unit1]
Node        > [interface]
Section     > (interface (type (TkNewType (TTestConstructor (TkClass class TkClassParents public (constructor (TkFunctionName Create) TkFunctionArgs))))))
Node        > [type]
Node        > [TkNewType]
Node        > [TTestConstructor]
Node        > [TkClass]
Node        > [class]
Node        > [TkClassParents]
Node        > [public]
Node        > [constructor]
Constructor > (constructor (TkFunctionName Create) TkFunctionArgs)
Node        > [TkFunctionName]
Node        > [Create]
Node        > [TkFunctionArgs]
Node        > [implementation]
Section     > (implementation (constructor (TkFunctionName TTestConstructor . Create) TkFunctionArgs) (begin inherited ; Writeln ( 'do something' ) ; end))
Node        > [constructor]
Constructor > (constructor (TkFunctionName TTestConstructor . Create) TkFunctionArgs)
Node        > [TkFunctionName]
Node        > [TTestConstructor]
Node        > [.]
Node        > [Create]
Node        > [TkFunctionArgs]
Node        > [begin]
Node        > [inherited]
Node        > [;]
Node        > [Writeln]
Node        > [(]
Node        > ['do something']
Node        > [)]
Node        > [;]
Node        > [end]
Node        > [end]
Node        > [.]
```

I _could_ catch `TkNewType` nodes and query their children for constructors, but I suspect something for records will change when you will fix #20, so I will wait for that before continuing to work on this issue.

**Comment by fabriciocolombo** [comment]
Yes, it will change after #20. 
I'm adding a new token TkRecord, but the change broke some tests, and i am correcting them.

**Comment by fabriciocolombo** [comment]
I need release the next version tomorrow, so i fixed this issue myself.
Thank you for your investigation and effort.
